securedrop/source_app/session_manager.py

```diff
@@ -0,0 +1,106 @@
+from typing import TYPE_CHECKING
+
+from flask import session, g
+from werkzeug.contrib.cache import SimpleCache
+
+from sdconfig import config
+from source_user import SourceUser
+
+if TYPE_CHECKING:
+    from passphrases import DicewarePassphrase
+
+
+class _InvalidUserSession(Exception):
+    pass
+
+
+class UserNotLoggedIn(_InvalidUserSession):
+    pass
+
+
+class UserSessionExpired(_InvalidUserSession):
+    pass
+
+
+class UserHasBeenDeleted(_InvalidUserSession):
+    pass
+
+
+class SessionManager:
+    _CACHE = SimpleCache()  # A mapping of passphrase -> SourceUser
+    _KEY_IN_SESSION_COOKIE = "codename"  # The key in flask.session for the user's passphrase
+
+    @classmethod
+    def log_user_in(cls, user: SourceUser, user_passphrase: "DicewarePassphrase") -> None:
+        # Save the passphrase in the user's session cookie
+        session[cls._KEY_IN_SESSION_COOKIE] = user_passphrase
```

Comment:
This follows the same logic as the current code base, which is to put the user's passphrase in their session cookie (which is encrypted). I don't know if this is useful or not, but we could replace the passphrase in the cookie with a randomly-generated session ID.